src/storage/src/read/windowed.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use arrow::compute::SortOptions;
+use arrow::row::{RowConverter, SortField};
+use datatypes::data_type::DataType;
+use datatypes::prelude::VectorRef;
+use datatypes::vectors::{Helper, MutableVector};
+use snafu::ResultExt;
+
+use crate::error;
+use crate::read::{Batch, BatchReader};
+use crate::schema::ProjectedSchemaRef;
+
+/// [WindowedReader] provides a windowed record batch reader that scans all rows within a window
+/// at a time and sort these rows ordered in `[<timestamp>, <PK>]` order.
+pub struct WindowedReader<R> {
+    /// Schema to read
+    pub schema: ProjectedSchemaRef,
+    /// Each reader reads a slice of time window
+    pub readers: Vec<R>,
+}
+
+impl<R> WindowedReader<R> {
+    pub fn new(schema: ProjectedSchemaRef, readers: Vec<R>) -> Self {
+        Self { schema, readers }
+    }
+}
+
+#[async_trait::async_trait]
+impl<R> BatchReader for WindowedReader<R>
+where
+    R: BatchReader,
+{
+    async fn next_batch(&mut self) -> error::Result<Option<Batch>> {
```

Comment:
Import crate::error::Result?
```suggestion
    async fn next_batch(&mut self) -> Result<Option<Batch>> {
```